cpp/src/tsne/bh_kernels.h

```diff
@@ -225,6 +184,10 @@ TreeBuildingKernel( /* int *restrict errd, */
   // iterate over all bodies assigned to thread
   while (i < N)
   {
+    if (++limiter > NNODES) {
+      break;
```

Comment:
Re: Everything related to `limiter` (the next several `break`s and `continue`s) - what does this address? (Crash, early bail performance optimization?)

Reply:
Oh yes, this was the primary fix for out of bounds memory accesses. Essentially, the loop could have executed forever, causing weird memory issues / never ending loop issues etc. A limiter is used to monitor this (as an easy fix) to force the loop to quit after all data points (ie NNODES) have been accessed. Since the loop should only have to ever check at most NNODES (ie the last layer of the KD-Tree) once.

Reply:
cc @drobison00

Reply:
Michael and I did a pretty thorough dive on the root cause of the lock ups. I'll have a more detailed description on Monday, but I don't think this resolves the problem. This code still has the potential to inject loops into the quad tree.